# Lab2: Multi-class Customer Segmentation

Target Variable: Segmentation

Algorithms:

    Decision Tree

    Logistic Regression

    Neural Network

## Experiment Goal

Students will build classification models to predict the customer segment of each individual using demographic and behavioral features. They will then compare model performance and explain which algorithm is most suitable for this dataset.

## Learning Objectives

1. By the end of the experiment, students should be able to:

1. Prepare data for classification tasks

1. Split data into training and testing sets

1. Train three different classifiers

1. Evaluate models using multiple metrics

1. Compare algorithm performance and interpret results

1. Explain advantages and disadvantages of each method

## Background of the Task

The dataset contains customer information such as:

    Gender

    Age

    Ever_Married

    Profession

    Work_Experience

    Spending_Score

    Family_Size

    Var_1

The prediction target is:

    Segmentation: A, B, C, or D

This is a multi-class classification problem.

## Experimental Requirements

Students must complete the following tasks:

### Part 1. Data Preparation

Students should:

    Load the dataset

    Inspect shape, column names, and data types

    Handle missing values

    Remove duplicates if necessary

    Encode categorical variables

    Scale features where appropriate

    Separate features X and target y



## Import Libraries

In [1]:
import warnings
warnings.filterwarnings('ignore')

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.preprocessing import LabelEncoder, StandardScaler

pd.set_option('display.max_columns', None)
sns.set_theme(style='whitegrid')

## Load dataset

In [ ]:
file_path = "Train.csv"
df = pd.read_csv(file_path)

for col in df.columns:
    if df[col] not in [ "Gender",   " Age",    "Ever_Married",    "Profession",
    "Work_Experience",

    "Spending_Score",

    "Family_Size",

    "Var_1"]
df = df.drop(columns=['ID'])

print("Dataset load successful!")
print("Dataset shape:", df.shape)

Dataset load successful!
Dataset shape: (10695, 11)


## Inspect shape, column names, and data types


In [3]:
df.head() #查看数据集的前几行,默认显示前5行

,ID,Gender,Ever_Married,Age,Graduated,Profession,Work_Experience,Spending_Score,Family_Size,Var_1,Segmentation
0,462809.0,Male,No,22,No,Healthcare,1.0,Low,4.0,Cat_4,D
1,462643.0,Female,Yes,38,Yes,Engineer,NaN,Average,3.0,Cat_4,A
2,466315.0,Female,Yes,67,Yes,Engineer,1.0,Low,1.0,Cat_6,B
3,461735.0,Male,Yes,67,Yes,Lawyer,0.0,High,2.0,Cat_6,B
4,462669.0,Female,Yes,40,Yes,Entertainment,NaN,High,6.0,Cat_6,A


In [4]:
df.info() #查看数据集的基本信息,包括每列的数据类型和非空值数量

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10695 entries, 0 to 10694
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               8068 non-null   float64
 1   Gender           10695 non-null  object 
 2   Ever_Married     10505 non-null  object 
 3   Age              10695 non-null  int64  
 4   Graduated        10593 non-null  object 
 5   Profession       10533 non-null  object 
 6   Work_Experience  9597 non-null   float64
 7   Spending_Score   10695 non-null  object 
 8   Family_Size      10247 non-null  float64
 9   Var_1            10587 non-null  object 
 10  Segmentation     10695 non-null  object 
dtypes: float64(3), int64(1), object(7)
memory usage: 919.2+ KB


In [11]:
df.describe(include='all').T
# 查看数据集的统计信息,包括数值型和分类型数据的统计描述
# include='all'参数表示同时显示数值型和分类型数据的统计信息
# .T表示转置,使得每列的统计信息显示在一行,更易于阅读

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
ID,8068.0,NaN,NaN,NaN,463479.214551,2595.381232,458982.0,461240.75,463472.5,465744.25,467974.0
Gender,10695,2,Male,5841,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Ever_Married,10505,2,Yes,6163,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Age,10695.0,NaN,NaN,NaN,43.511828,16.774158,18.0,30.0,41.0,53.0,89.0
Graduated,10593,2,Yes,6570,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Profession,10533,9,Artist,3318,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Work_Experience,9597.0,NaN,NaN,NaN,2.619777,3.39079,0.0,0.0,1.0,4.0,14.0
Spending_Score,10695,3,Low,6494,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Family_Size,10247.0,NaN,NaN,NaN,2.844052,1.536427,1.0,2.0,3.0,4.0,9.0
Var_1,10587,7,Cat_6,6910,NaN,NaN,NaN,NaN,NaN,NaN,NaN


**Observation**:

  Dataset Overview                                                                                                                                                                                                   
                                                                                                                                                                                                                       - 10,695 total records with 11 columns                                                                                                                                                                               - ID: Sequential identifier (458,982 to 467,974)                                                                                                                                                                   
                                                                                                                                                                                                                     
  Demographic Profile

  - Gender: Male majority (5,841, ~54.6%)
  - Age: Mean 43.5 years, range 18-89, median 41
  - Marital Status: Majority married (6,163 of 10,505, ~58.7%)
  - Education: Majority graduated (6,570 of 10,593, ~62%)

  Professional Characteristics

  - Work Experience: Mean 2.6 years, range 0-14, median 1 year
  - Profession: 9 unique types, most common "Artist" (3,318 of 10,533)

  Behavioral & Lifestyle

  - Spending Score: Majority "Low" (6,494 of 10,695)
  - Family Size: Mean 2.84, range 1-9, median 3
  - Var_1: 7 categories, most common "Cat_6" (6,910 of 10,587)

  Target Segmentation

  - 4 customer segments, most common is Segment D (3,027 of 10,695)

  Data Quality Notes

  - Missing values present (different counts across columns)
  - ID appears sequential with no gaps
  - Numerical features show normal distributions (Age, Work_Experience, Family_Size)

  Key Patterns

  1. Relatively young professional population (median age 41)
  2. Conservative spending behavior dominates
  3. Low average work experience (median 1 year)
  4. Small to medium family sizes typical
  5. Segment D is the largest customer group

## Handle missing value and duplicates

In [13]:
missing_count = df.isnull().sum() # 计算每列的缺失值数量
missing_percentage = ((missing_count / len(df)) * 100).round(2) # 计算每列的缺失值百分比

missing_table = pd.DataFrame(
    {
        'Missing Count': missing_count,
        'Missing Percentage (%)': missing_percentage
    }
).sort_values(by='Missing Percentage (%)', ascending=False) # 按照缺失值百分比降序排序

missing_table

,Missing Count,Missing Percentage (%)
ID,2627,24.56
Work_Experience,1098,10.27
Family_Size,448,4.19
Ever_Married,190,1.78
Profession,162,1.51
Var_1,108,1.01
Graduated,102,0.95
Gender,0,0.00
Age,0,0.00
Spending_Score,0,0.00


### Part 2. Train-Test Split

Students divide the dataset into:

    Training set: 80%

    Testing set: 20%

Use a fixed random seed such as random_state=42.

## Models to Be Built
### Model 1: Decision Tree

Students train a decision tree classifier.

Suggested settings:

    criterion="gini" or criterion="entropy"

    control tree depth to reduce overfitting

Students should observe:

    whether the tree overfits

    whether the model is easy to interpret

### Model 2: Logistic Regression

Students train a logistic regression classifier for multi-class prediction.

Suggested settings:

    use proper solver such as lbfgs

    increase max_iter if needed

    standardize features before training

Students should observe:

    whether linear decision boundaries are sufficient

    whether performance is stable

### Model 3: Neural Network

Students train a simple feedforward neural network.

Suggested settings:

    input layer = number of features

    1–2 hidden layers

    output layer = 4 classes

    activation: ReLU for hidden layers，softmax for output layer

Students may implement:

    MLPClassifier from scikit-learn or a Keras/TensorFlow model if allowed

Students should observe:

    whether the neural network improves performance

    whether training is more complex than the other methods

## Evaluation Metrics

Students must evaluate all three models using the same test set.

Required metrics:

    Accuracy

    Precision

    Recall

    F1-score

    Confusion Matrix

For multi-class classification, students should report:

    macro average

    weighted average

This helps them compare overall performance more fairly.

## Comparison Task

Students must compare the three algorithms from the following aspects:

    (1) Predictive Performance
        Which model has the highest:
        accuracy
        F1-score
        balanced performance across all classes

    (2) Overfitting Risk
        Did the decision tree overfit?
        Did the neural network show unstable behavior?
        Was logistic regression more stable?

    (3) Interpretability
        Decision tree is usually easiest to explain
        Logistic regression provides linear relationships
        Neural network is often hardest to interpret

    (4) Computational Complexity
        Which model trained fastest?
        Which model required the most preprocessing?
        Which model was hardest to tune?

    (5) Suitability for This Dataset
        Students should discuss which algorithm is best for this specific dataset, not in general.